pkg/controller/cronjob/cronjob_controller2.go

```diff
@@ -0,0 +1,580 @@
+			jm.enqueueController(cronJob)
+		}
+	}
+
+	// If it has a ControllerRef, that's all that matters.
+	if curControllerRef != nil {
+		cronJob := jm.resolveControllerRef(curJob.Namespace, curControllerRef)
+		if cronJob == nil {
+			return
+		}
+		jm.enqueueController(cronJob)
+		return
+	}
+}
+
+func (jm *Controller2) deleteJob(obj interface{}) {
+	job, ok := obj.(*batchv1.Job)
+
+	// When a delete is dropped, the relist will notice a job in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		job, ok = tombstone.Obj.(*batchv1.Job)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a ReplicaSet %#v", obj))
+			return
+		}
+	}
+
+	controllerRef := metav1.GetControllerOf(job)
+	if controllerRef == nil {
+		// No controller should care about orphans being deleted.
+		return
+	}
+	cronJob := jm.resolveControllerRef(job.Namespace, controllerRef)
+	if cronJob == nil {
+		return
+	}
+	jm.enqueueController(cronJob)
+}
+
+func (jm *Controller2) enqueueController(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.Add(key)
+}
+
+func (jm *Controller2) enqueueControllerWithTime(obj interface{}, t time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.AddAfter(key, t)
+}
+
+func (jm *Controller2) updateCronJob(old interface{}, curr interface{}) {
+	//  the big cases that needs to be handled
+	//  1. change in CJ.spec.schedule
+	//  2. change in CJ.spec.JobTemplate
+	//  other CJ.spec params seem to be dealt with by the controller in sync loop on updates
+	oldCJ, okOld := old.(*batchv1beta1.CronJob)
+	newCJ, okNew := curr.(*batchv1beta1.CronJob)
+
+	if !okOld || !okNew {
+		// typecasting of one failed, handle this better, may be log entry
+		return
+	}
+	// if the change in schedule results in next requeue having to be sooner than it already was,
+	// it will be handled here by the queue. If the next requeue is further than previous schedule,
+	// the sync loop will essentially be a no-op for the already queued key with old schedule.
+	if oldCJ.Spec.Schedule != newCJ.Spec.Schedule {
+		// schedule changed, change the requeue time
+		sched, err := cron.ParseStandard(newCJ.Spec.Schedule)
+		if err != nil {
+			// this is likely a user error in defining the spec value
+			// we should log the error and not reconcile this cronjob until an update to spec
+			klog.Errorf("unparseable schedule: %s : %s", newCJ.Spec.Schedule, err)
```

Comment:
Theoretically this should never pass validation, but a check here is good :+1:

Reply:
But try not to use `klog.Errorf` only in rare situation, this can be at most `klog.V(2)`

Reply:
+1 addressed